language-support/ts/codegen/src/TsCodeGenMain.hs

```diff
@@ -388,12 +388,13 @@ genDefDataType curPkgId conName mod tpls def =
                             ] ++
                             ["};"]
                         associatedTypes =
-                          let tT = conName
-                              tK = tT <> ".Key"
+                          let keyDef = maybe "unknown" (fst . genType (moduleName mod) . tplKeyType) (tplKey tpl)
+                              tT = conName
+                              tK = if keyDef == "unknown" then keyDef else tT <> ".Key"
```

Comment:
Shouldn’t this be `undefined`instead of `unknown`? Also is there any downside to simply always setting this to `keyDef` instead of going via the `.Key` indirection?

Reply:
> Shouldn’t this be `undefined`instead of `unknown`?

I agree this is formally more correct. Switching to `undefined`.

> Also is there any downside to simply always setting this to `keyDef` instead of going via the `.Key` indirection?

No technical downside. The `.Key` approach leads to prettier generated code is all.